packages/core/src/schema/schema.ts

```diff
@@ -1,341 +1,212 @@
-
-// Find the name and base type of a field definition,
-// handling any wrapper types (NON_NULL_TYPE and LIST_TYPE).
-const unwrapFieldDefinition = (field: FieldDefinitionNode) => {
-  const fieldName = field.name.value;
-  let fieldType = field.type;
-  let isNotNull = false;
-  let isList = false;
-  let isListElementNotNull = false;
-
-  // First check if the field is non-null and unwrap it.
-  if (fieldType.kind === Kind.NON_NULL_TYPE) {
-    isNotNull = true;
-    fieldType = fieldType.type;
-  }
-
-  // Then check if the field is a list and unwrap it.
-  if (fieldType.kind === Kind.LIST_TYPE) {
-    isList = true;
-    fieldType = fieldType.type;
-
-    // Now check if the list element type is non-null
-    if (fieldType.kind === Kind.NON_NULL_TYPE) {
-      isListElementNotNull = true;
-      fieldType = fieldType.type;
-    }
-  }
-
-  if (fieldType.kind === Kind.LIST_TYPE) {
-    throw new Error(
-      `Invalid field "${fieldName}": nested lists are not supported`
-    );
-  }
+export const createTable = <TTableName extends string>(
+  name: TTableName
+): IT<TTableName, {}> => {
+  const table = { name, columns: {} } as const;
 
   return {
-    fieldName,
-    fieldTypeName: fieldType.name.value,
-    isNotNull,
-    isList,
-    isListElementNotNull,
+    table,
+    addColumn: <
+      TTables extends readonly Table[],
+      TName extends string,
+      TType extends Scalar,
+      TReferneces extends "id" extends TName
+        ? never
+        : `${(TTables & Table[])[number]["name"] & string}.id`,
+      TOptional extends "id" extends TName ? false : boolean = false,
+      TList extends "id" extends TName ? false : boolean = false
+    >(
+      name: TName,
+      type: TType,
+      modifiers?: {
+        references?: TReferneces;
+        optional?: TOptional;
+        list?: TList;
+      }
+    ): IT<
+      TTableName,
+      Record<TName, Column<TTables, TType, TReferneces, TOptional, TList>>
+    > => addColumn(table, name, type, modifiers),
   };
 };
 
-// Find all types in the schema that are marked with the @entity directive.
-const getEntityTypes = (schema: GraphQLSchema) => {
-  const entities = Object.values(schema.getTypeMap())
-    .filter((type): type is GraphQLObjectType => {
-      return type.astNode?.kind === Kind.OBJECT_TYPE_DEFINITION;
-    })
-    .filter((type) => {
-      return !!type.astNode?.directives?.find(
-        (directive) => directive.name.value === "entity"
-      );
+/**
+ * Type inference and runtime validation
+ */
+export const createSchema = <
+  TSchema extends readonly IT<
+    string,
+    { id: Column<Table[], ID, never, false, false> } & Record<
+      string,
+      Column<Table[], Scalar, unknown, boolean, boolean>
+    >
+  >[]
+>(
+  schema: TSchema
+): {
+  entities: { [key in keyof TSchema]: TSchema[key]["table"] };
+} => {
+  const tables = schema.map((it) => it.table);
+
+  tables.forEach((t) => {
```

Comment:
Love this runtime validation. At some point we'll need to test that the errors are propagating properly when making changes with the dev server active. Lmk if you want help with this, can point you in the right direction.